File: src/main/java/com/jcloisterzone/game/phase/RewindActionContainer.java
package com.jcloisterzone.game.phase;

/**
 *  Rewind Actions container for option when Towr Ransom Pay, possibility to revert Action Phase
 *  ActionPhase can be skipped empty Actions
 *  PhantomPhase can be skipped when no Phantom avaialble (captured by tower)
 */
public class RewindActionContainer {

    private Phase actionPhase;
    private Phase phantomPhase;

    public RewindActionContainer() {
    	
    }

	public RewindActionContainer(Phase actionPhase, Phase phantomPhase) {
        this.actionPhase = actionPhase;
        this.phantomPhase = phantomPhase;
    }

    public RewindActionContainer setActionPhase(Phase actionPhase) {
        this.actionPhase = actionPhase;
        return this;
    }

    public Phase getActionPhase() {
    	return actionPhase;
    }

    public RewindActionContainer setPhantomPhase(Phase phantomPhase) {
        this.phantomPhase = phantomPhase;
        return this;
    }

    public Phase getPhantomPhase() {
    	return phantomPhase;
    }

}

File: src/main/ts/com/jcloisterzone/game/phase/RewindActionContainer.ts
import type { Phase } from "./Phase.js";

/**
 * Holds references to the ActionPhase and PhantomPhase so that a Tower "random pay"
 * (PAY_RANSOM) performed after the wood action — but before it has been confirmed or a
 * post-wood follow-up has begun — can rewind the turn back to those phases.
 */
export class RewindActionContainer {
  private actionPhase: Phase | null;
  private phantomPhase: Phase | null;

  constructor(actionPhase: Phase | null = null, phantomPhase: Phase | null = null) {
    this.actionPhase = actionPhase;
    this.phantomPhase = phantomPhase;
  }

  setActionPhase(actionPhase: Phase | null): RewindActionContainer {
    this.actionPhase = actionPhase;
    return this;
  }

  getActionPhase(): Phase | null {
    return this.actionPhase;
  }

  setPhantomPhase(phantomPhase: Phase | null): RewindActionContainer {
    this.phantomPhase = phantomPhase;
    return this;
  }

  getPhantomPhase(): Phase | null {
    return this.phantomPhase;
  }
}
